src/polyfem/assembler/GenericProblem.cpp

```diff
@@ -80,6 +86,130 @@ namespace polyfem
+				for (size_t i = 0; i < points_.size()-1; ++i) {
+					if (t-points_[points_.size()-1] >= points_[i] && t-points_[points_.size()-1] < points_[i+1]) {
+						return values_[i] + offset_;
+					}
+				}
+			}
+
+			return 0;
+		}
+
+		double PiecewiseLinearInterpolation::eval(const double t) const
+		{
+			if (t < points_[0])
+				return 0;
+			for (size_t i = 0; i < points_.size()-1; ++i) {
+				if (t >= points_[i] && t < points_[i+1]) {
+					double val = (values_[i+1]-values_[i]) * ((t-points_[i])/(points_[i+1]-points_[i])) + values_[i];
+					return val;
+				}
+			}
+			if (extend_ == 0)
+				return values_[values_.size()-1];
+			else if (extend_ == 1)
+				return values_[values_.size()-1]*t;
+			else if (extend_ == 2) {
+				for (size_t i = 0; i < points_.size()-1; ++i) {
+					if (t-points_[points_.size()-1] >= points_[i] && t-points_[points_.size()-1] < points_[i+1]) {
+						double val = (values_[i+1]-values_[i]) * ((t-points_[i])/(points_[i+1]-points_[i])) + values_[i];
+						return val;
+					}
+				}
+			}
+			else if (extend_ == 3) {
+				for (size_t i = 0; i < points_.size()-1; ++i) {
+					if (t-points_[points_.size()-1] >= points_[i] && t-points_[points_.size()-1] < points_[i+1]) {
+						double val = (values_[i+1]-values_[i]) * ((t-points_[i])/(points_[i+1]-points_[i])) + values_[i];
+						return val + offset_;
+					}
+				}
+			}
+
+			return 0;
+		}
+
+
+		void PiecewiseConstantInterpolation::init(const json &params)
+		{
+			if (!params["points"].is_array())
+				throw "Points must be an array";
+			if (!params["values"].is_array())
+				throw "Values must be an array";
+			
+			points_.reserve(params["points"].size());
+			values_.reserve(params["values"].size());
+
+			for (int i = 0; i < params["points"].size(); ++i) {
+				points_[i] = params["points"][i];
+				values_[i] = params["values"][i];
+			}
+
+			if (params.contains("extend")) {
+				if (params["extend"] == "constant")
+					extend_ = 0;
+				else if (params["extend"] == "extrapolate")
+					extend_ = 1;
+				else if (params["extend"] == "repeat")
+					extend_ = 2;
+				else if (params["extend"] == "repeat_offset")
+					extend_ = 3;
+			}
+			else
+				extend_ = -1;
+		}
+
+		void PiecewiseLinearInterpolation::init(const json &params)
+		{
+			if (!params["points"].is_array())
+				throw "Points must be an array";
+			if (!params["values"].is_array())
+				throw "Values must be an array";
+			
+			points_.reserve(params["points"].size());
+			values_.reserve(params["values"].size());
+
+			for (int i = 0; i < params["points"].size(); ++i) {
+				points_[i] = params["points"][i];
+				values_[i] = params["values"][i];
+			}
+
+			if (params.contains("extend")) {
+				if (params["extend"] == "constant")
+					extend_ = 0;
+				else if (params["extend"] == "extrapolate")
+					extend_ = 1;
+				else if (params["extend"] == "repeat")
+					extend_ = 2;
+				else if (params["extend"] == "repeat_offset")
+					extend_ = 3;
+			}
+			else
+				extend_ = -1;
```

Comment:
Does FEBio define a default? We should use that same one. or else throw and error here.

Reply:
Default is constant in FEBio: https://help.febio.org/FebioUser/FEBio_um_3-4-Subsection-3.17.1.html